syntax = "proto3";

package envoy.extensions.filters.http.payload_validator.v3;

import "envoy/config/core/v3/base.proto";
import "envoy/type/http_status.proto";

import "google/protobuf/wrappers.proto";

import "udpa/annotations/status.proto";
import "validate/validate.proto";

option java_package = "io.envoyproxy.envoy.extensions.filters.http.payload_validator.v3";
option java_outer_classname = "PayloadValidatorProto";
option java_multiple_files = true;
option go_package = "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/http/payload_validator/v3;payload_validatorv3";
option (udpa.annotations.file_status).package_version_status = ACTIVE;

// [#protodoc-title: Payload Validator]

message PayloadDescription {
    string type = 1;
    string schema = 2;
}

enum ParameterLocation {
  QUERY = 0;
  PATH = 1;
}

message Parameter {
    string name = 1 [(validate.rules).string = {min_len: 1}];
    ParameterLocation in = 2 [(validate.rules).enum = {defined_only: true}];
    google.protobuf.BoolValue required = 3;
    string schema = 4 [(validate.rules).string = {min_len: 5}];
}

message ResponseDescription {
    type.HttpStatus  http_status = 1 [(validate.rules).message = {required: true}];
    PayloadDescription response_body = 2;
}

message TransactionDescription {
       config.core.v3.RequestMethod method = 1 [(validate.rules).enum = {defined_only: true}];
       repeated Parameter parameters = 2;
       PayloadDescription request_body = 3;
       repeated ResponseDescription responses = 4;
}

message PathTransactions {
  string path = 1 [(validate.rules).string = {min_len: 1}];
  repeated TransactionDescription operations = 2 [(validate.rules).repeated = {min_items: 1}];
}


// Payload Validator filter config.
message PayloadValidator {
  string stat_prefix = 1 [(validate.rules).string = {min_len: 1}];
  google.protobuf.UInt32Value max_size = 2;
  repeated PathTransactions paths = 3 [(validate.rules).repeated = {min_items: 1}];
}
